std.os.uefi.protocol: ziggify function signatures

dotcarmen opened the pull request:
A lot of the function signatures in `std.os.uefi.protocol` aren't very ziggy. This PR modifies a significant number of the namespace's functions to align more with verbose Zig code.

- Some parameters changed from `*` to `*const` and vice-versa as appropriate
- Some pairs of parameters that represent slices were consolidated into a single slice parameter
- Status errors associated with each function according to the [UEFI Specification](https://uefi.org/sites/default/files/resources/UEFI_Spec_Final_2.11.pdf) are handled specifically, resulting in narrow error unions being returned from each function
  - unexpected status codes result in a new call to `uefi.unexpectedError`, which is akin to `posix.unexpectedError`
- As appropriate, return values from various functions are returned directly when the underlying function's return value is `.success`, rather than requiring an output pointer parameter

note: most of `std.os.uefi` is untested, and that's not really changing here. I'm mostly focused on making sure there are no compiler errors. Any further problems will require a future PR to address.

linusg commented on lib/std/os/uefi/protocol/device_path.zig:
```diff
@@ -42,7 +42,7 @@ pub const DevicePath = extern struct {
 
     /// Creates a file device path from the existing device path and a file path.
-    pub fn create_file_device_path(self: *DevicePath, allocator: Allocator, path: [:0]align(1) const u16) !*DevicePath {
+    pub fn create_file_device_path(self: *const DevicePath, allocator: Allocator, path: []const u16) !*DevicePath {
```
Function name should be camel case

linusg commented on lib/std/os/uefi/protocol/edid.zig:
```diff
@@ -44,10 +45,19 @@ pub const Override = extern struct {
+        if (size == 0 or ptr == null)
+            return null;
+
+        return ptr[0..size];
```
I'd assume this doesn't work as `ptr` needs to be null-unwrapped for slicing

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,95 +57,161 @@ pub const File = extern struct {
+    pub fn delete(self: *File) bool {
+        switch (self._delete(self)) {
+            .success => return true,
+            .warn_delete_failure => return false,
```
That seems like a good solution. Once we find an API that can return both a value and a warning it'll get awkward though.

linusg commented on lib/std/os/uefi/protocol/edid.zig:
```diff
@@ -44,10 +45,19 @@ pub const Override = extern struct {
-    ) Status {
-        return self._get_edid(self, handle, attributes, edid_size, edid);
+    ) !?[]u8 {
+        var size: usize = 0;
+        var ptr: ?[*]u8 = null;
```
I think we can use `= undefined` for these

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,95 +57,161 @@ pub const File = extern struct {
+        self: *const File,
+        information_type: *align(8) const Guid,
+        buffer: []u8,
+    ) !usize {
```
Probably more idiomatic to return a slice here like in `std.fmt.bufPrint()`

linusg commented on lib/std/os/uefi/protocol/graphics_output.zig:
```diff
@@ -3,16 +3,23 @@ const uefi = std.os.uefi;
+            .success => return info,
+            .device_error => return Error.DeviceError,
+            .invalid_parameter => return Error.InvalidParameter,
+            else => |err| uefi.unexpectedError(err.err()),
```
```suggestion
            else => |err| uefi.unexpectedError(err),
```

dotcarmen commented on lib/std/os/uefi.zig:
```diff
@@ -247,3 +247,7 @@ test {
 }
+
+pub fn unexpectedError(err: Status) noreturn {
+    std.debug.panic("unexpected error: {any}", .{err});
```
I didn't want to diverge meaningfully from `Status.Error` being the return value, but as I've been updating my own codebase to use this updated API, I agree it's not very sensible :) I'll follow `posix.unexpectedErrno` better in the next commit

dotcarmen replied:
hm, notably this does break a helper function i have to convert `uefi.Status.Error -> uefi.Status`, because it's apparently not possible to narrow `uefi.Status.Error || error{Unexpected}` back to `uefi.Status.Error`, even with `inline` switching...

not the worst thing in the world, but certainly inconvenient

linusg replied:
```zig
const err: (uefi.Status.Error || error{Unexpected}) = error.DeviceError;
const casted: uefi.Status.Error = @errorCast(err);
```

I think this should work, no?

dotcarmen replied:
lol yeah i forgot about `@errorCast` thanks

dotcarmen commented on lib/std/os/uefi/protocol/ip6.zig:
```diff
@@ -7,61 +7,245 @@ const MacAddress = uefi.MacAddress;
     /// Polls for incoming data packets and processes outgoing data packets.
-    pub fn poll(self: *const Ip6) Status {
-        return self._poll(self);
+    pub fn poll(self: *Ip6) PollError!void {
```
i feel like this should return `PollError!bool`, where `.success` is `true` and `.not_ready` is `false` - wdyt? according to the [spec](https://uefi.org/specs/UEFI/2.10/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-ip6-protocol-poll), poll returns `.not_ready` when no data is processed, as opposed to `.success` when data _is_ processed

dotcarmen commented on lib/std/os/uefi/protocol/ip6_config.zig:
```diff
@@ -4,27 +4,99 @@ const Guid = uefi.Guid;
+
+    pub fn setData(
+        self: *const Ip6Config,
+        data_type: DataType,
+        data_size: usize,
+        data: *const anyopaque,
```
the data type should probably be a `union (enum)` eventually, but i'm concerned about how that would affect `getData` where the out pointer is the field defined by `data_type` from the union

dotcarmen replied:
did it

dotcarmen commented on lib/std/os/uefi/protocol/ip6_service_binding.zig:
```diff
@@ -4,17 +4,47 @@ const Handle = uefi.Handle;
+    pub const CreateChildError = uefi.UnexpectedError || error{
+        InvalidParameter,
+        OutOfResources,
+    } || Error; // TODO: according to the spec, _any other_ status is returnable?
```
i only did a little bit of google, but couldn't figure out what `Other` means in the [spec](https://uefi.org/specs/UEFI/2.10/11_Protocols_UEFI_Driver_Model.html#efi-service-binding-protocol-createchild). i have a feeling it's referring to warnings, but i'm wary of assuming such and i think i'd rather let somebody who encounters this file their own bug report/pr to fix...

dotcarmen commented on lib/std/os/uefi/protocol/managed_network.zig:
```diff
@@ -8,58 +8,209 @@ const Time = uefi.Time;
+        var data: GetModeDataData = undefined;
+        switch (self._get_mode_data(self, &data.mnp_config, &data.snp_mode)) {
+            .success => return data,
+            // .invalid_parameter => return Error.InvalidParameter,
+            // .unsupported => return Error.Unsupported,
+            // .not_started => return Error.NotStarted,
+            else => |status| {
+                try status.err();
+                // TODO: only warnings get here I think?
+                return uefi.unexpectedStatus(status);
```
depending on how we decide on approaching `Other` above, i'll change the logic here

dotcarmen commented:
for warnings, i'm considering a global `pub var warning: Status = .success;` in `uefi.zig`, which the user can check for a value. then, `uefi.unexpectedStatus` can return `UnexpectedError!void` and callers can return `.success` value.

alternatively, I can make a `pub fn MaybeWarned(T: type) type` which has an optional `warning: ?Status` field in addition to `result: T`. thoughts?

dotcarmen commented on lib/std/os/uefi/protocol/graphics_output.zig:
```diff
@@ -3,26 +3,84 @@ const uefi = std.os.uefi;
+            else => |status| uefi.unexpectedStatus(status),
+        }
+
+        if (size_of_info != @sizeOf(Mode.Info))
+            return error.Unexpected
```
also thoughts on this pattern? the 3rd parameter is for informing the caller of the implementation's `sizeOf Mode.Info` even though the struct is well-defined according to the spec...

this also applies to [`SimpleNetwork.statistics`](https://uefi.org/specs/UEFI/2.10/24_Network_Protocols_SNP_PXE_BIS.html#efi-simple-network-statistics)

dotcarmen replied:
i think it makes sense to assume/return `Mode.Info` and if the user wants/needs to override the behavior, they can call the underlying fn directly or open a pr

linusg replied:
> even though the struct is well-defined according to the spec...

I assume it's expected to change across spec versions:

> The size of the Info structure should never be assumed and the value of SizeOfInfo is the only valid way to know the size of Info.

Assuming that they would only add fields over time I think we can disregard the size we're told and simply return our definition of `Mode.Info`.

dotcarmen replied:
that makes sense, thanks

dotcarmen commented:
as mentioned in the last commit message, i've finished the initial refactorings :D I don't have time for a self-review right now, I'll get to it later today (I'm traveling in Japan for another couple days)

dotcarmen commented on lib/std/os/uefi/protocol/simple_network.zig:
```diff
@@ -4,88 +4,354 @@ const Event = uefi.Event;
     /// Reads the current interrupt status and recycled transmit buffer status from a network interface.
-    pub fn getStatus(self: *const SimpleNetwork, interrupt_status: *InterruptStatus, tx_buf: ?*?[*]u8) Status {
-        return self._get_status(self, interrupt_status, tx_buf);
+    pub fn getStatus(
+        self: *SimpleNetwork,
+        interrupt_status: ?*InterruptStatus,
+        tx_buf: ?*?[*]u8,
+    ) GetStatusError!void {
```
lol there's 4 different ways to return values here, so i just left it as is...

dotcarmen commented on lib/std/os/uefi/protocol/simple_text_input.zig:
i feel like this implementation is unnecessary, it's redundant to `simple_text_input_ex`, wdyt?

linusg replied:
Unless the UEFI spec completely drops the basic variant in favor of the extended one I'd keep both.

dotcarmen commented on lib/std/os/uefi/protocol/simple_text_output.zig:
```diff
@@ -3,63 +3,141 @@ const uefi = std.os.uefi;
-        return self._output_string(self, msg);
+    ///
+    /// Returns `true` if the string was successfully written, `false` if an unknown glyph was encountered.
+    pub fn outputString(self: *SimpleTextOutput, msg: [*:0]const u16) OutputStringError!bool {
+        switch (self._output_string(self, msg)) {
+            .success => return true,
+            .warn_unknown_glyph => return false,
```
another instance of a defined warning being possible, luckily this one is also easy...

dotcarmen commented on lib/std/os/uefi/protocol/simple_text_output.zig:
```diff
@@ -3,63 +3,141 @@ const uefi = std.os.uefi;
     /// Sets the background and foreground colors for the outputString() and clearScreen() functions.
-    pub fn setAttribute(self: *const SimpleTextOutput, attribute: usize) Status {
-        return self._set_attribute(self, attribute);
+    pub fn setAttribute(self: *SimpleTextOutput, attribute: usize) SetAttributeError!void {
```
this should also be a type

linusg reviewed (commented):
Great work, I really like where this is going 🙂

Please ping me when you consider this done and want a full review.

dotcarmen commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,95 +94,161 @@ pub const File = extern struct {
-        return self._get_info(self, information_type, buffer_size, buffer);
+    pub fn getInfo(
+        self: *const File,
+        information_type: *align(8) const Guid,
+        buffer: []u8,
```
i can do a similar change to `setInfo` if this is indeed better :)

dotcarmen commented on lib/std/os/uefi/protocol/graphics_output.zig:
```diff
@@ -3,26 +3,84 @@ const uefi = std.os.uefi;
     /// Returns information for an available graphics mode that the graphics device and the set of active video output devices supports.
-    pub fn queryMode(self: *const GraphicsOutput, mode: u32, size_of_info: *usize, info: **Mode.Info) Status {
-        return self._query_mode(self, mode, size_of_info, info);
+    pub fn queryMode(self: *const GraphicsOutput, mode: u32) QueryModeError!*Mode.Info {
```
nevermind, `mode` is a numeric id

dotcarmen commented on lib/std/os/uefi/protocol/hii_popup.zig:
```diff
@@ -4,15 +4,33 @@ const Guid = uefi.Guid;
+        style: PopupStyle,
+        popup_type: PopupType,
+        handle: hii.Handle,
+        msg: u16,
```
to be clear: the value of `msg` should be determined by using `EFI_STRING_PROTOCOL`

dotcarmen commented:
in addition to the specific requests for feedback i pinged you on, i think i'm ready for a full review :) @linusg

linusg commented on lib/std/os/uefi/protocol.zig:
```diff
@@ -23,16 +26,98 @@ pub const edid = @import("protocol/edid.zig");
+            switch (self._create_child(self, &handle)) {
+                .success => handle,
+                // .invalid_parameter => error.InvalidParameter,
+                // .out_of_resources => error.OutOfResources,
```
No commmented out code

linusg commented on lib/std/os/uefi/protocol.zig:
```diff
@@ -23,16 +26,98 @@ pub const edid = @import("protocol/edid.zig");
+                // .out_of_resources => error.OutOfResources,
+                else => |status| {
+                    try status.err();
+                    // TODO: only warnings get here???
```
Unactionable TODO

linusg commented on lib/std/os/uefi/protocol/edid.zig:
```diff
@@ -39,15 +40,25 @@ pub const Discovered = extern struct {
-        edid: *?[*]u8,
-    ) Status {
-        return self._get_edid(self, handle, attributes, edid_size, edid);
+    pub fn getEdid(self: *const Override, handle: Handle) GetEdidError!Edid {
```
I believe this should take a `*const Handle`:

> Note: the (EFI_HANDLE ) type of the “ChildHandle” parameter is an historical typing error in the UEFI specification. To match existent practice however, implementors and callers of the protocol are now expected to conform to the declaration of the parameter as written. That is, callers must pass the address of an EFI_HANDLE object as “ChildHandle”, and implementors must dereference “ChildHandle” for finding the EFI_HANDLE object.

dotcarmen replied:
ah you're right, i read that wrong last night. fixed the underlying function signature, and now the argument is passed in by taking the address of the *argument* instead of forcing the caller to take the address...

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,108 +94,195 @@ pub const File = extern struct {
-    pub fn setInfo(self: *const File, information_type: *align(8) const Guid, buffer_size: usize, buffer: [*]const u8) Status {
-        return self._set_info(self, information_type, buffer_size, buffer);
+    pub fn setInfo(
+        self: *const File,
```
Should probably be a mutable pointer

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,108 +94,195 @@ pub const File = extern struct {
+        create = 0x8000000000000000,
+    };
+
+    pub const Attributes = packed struct(u64) {
```
These are duplicated in `std/os/uefi.zig`, maybe we can use this as the source of truth

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,108 +94,195 @@ pub const File = extern struct {
+        // 0x0000000000000020
+        archive: bool = false,
+        // used exclusively for `valid_attr` as far as i can tell...
+        _flag: bool = false,
```
I don't think we need this as a separate field, or the `valid_attr` constant below.

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,108 +94,195 @@ pub const File = extern struct {
-    pub const efi_file_valid_attr: u64 = 0x0000000000000037;
+        // 0x0000000000000037
+        pub const valid_attr: Attributes = .{
+            .read_only = true,
+            .system = true,
+            ._flag = true,
```
This doesn't look correct, `0x37` is `0b110111`, i.e. all known fields but `reserved` set to true. But as mentioned I think this is unnecessary, this only seems to be used by UEFI implementations to check if the incoming flag matches this bitmask.

dotcarmen replied:
lol thanks for checking me on my math, i kept mixing decimal and hex in my head 😭 but yeah that makes sense

linusg commented on lib/std/os/uefi/protocol/file.zig:
```diff
@@ -40,108 +94,195 @@ pub const File = extern struct {
+    };
 
-    pub const efi_file_position_end_of_file: u64 = 0xffffffffffffffff;
+    const efi_file_position_end_of_file: u64 = 0xffffffffffffffff;
```
Drop `efi_file_` (see https://ziglang.org/documentation/master/#Avoid-Redundancy-in-Names)

dotcarmen commented:
ok fixed the compiler errors :) done pushing for the (japanese) night!

linusg commented on lib/std/os/uefi/protocol/graphics_output.zig:
```diff
@@ -3,26 +3,84 @@ const uefi = std.os.uefi;
-        return self._blt(self, blt_buffer, blt_operation, source_x, source_y, destination_x, destination_y, width, height, delta);
+    pub fn blt(
+        self: *GraphicsOutput,
+        blt_buffer: ?[*]BltPixel,
+        blt_operation: BltOperation,
+        source_x: usize,
+        source_y: usize,
+        destination_x: usize,
+        destination_y: usize,
+        width: usize,
+        height: usize,
+        delta: usize,
```
Seems fine, not necessary to diverge from the original API too much IMO.

linusg commented on lib/std/os/uefi/protocol/managed_network.zig:
```diff
@@ -8,58 +8,209 @@ const Time = uefi.Time;
+    _cancel: *const fn (*ManagedNetwork, ?*const CompletionToken) callconv(cc) Status,
+    _poll: *const fn (*ManagedNetwork) callconv(cc) Status,
+
+    // TODO: according to the spec, "Other" errors are also possible for some of these fns,
+    // but i think it's referring to warnings?
```
I don't think they refer to warnings, this seems to be dependent on the "MNP child driver" (whatever that is) - so we'll have to expect any UEFI error here.

dotcarmen commented:
addressed all feedback, ready for 🤞🏻 final review @linusg 

the only changes i made that were beyond the previous reviews' comments were 
dc2df7c and 478bae7 - if these are too weird, i can revert.

dotcarmen commented on lib/std/os/uefi/protocol/device_path.zig:
```diff
@@ -67,7 +73,7 @@ pub const DevicePath = extern struct {
         ptr[path.len] = 0;
 
-        var end = @as(*uefi.DevicePath.End.EndEntireDevicePath, @ptrCast(@as(*DevicePath, @ptrCast(new)).next().?));
+        var end = @as(*uefi.DevicePath.End.EndEntireDevicePath, @ptrCast(@constCast(@as(*DevicePath, @ptrCast(new)).next().?)));
```
i wanna make a follow-up pr to enable testing `std.os.uefi` in ci since that would've caught this...

dotcarmen replied:
https://github.com/ziglang/zig/pull/23286

linusg replied:
I don't think that change is sufficient for testing this, given Zig's lazy semantic analysis and no testing code would currently reference this.

dotcarmen commented:
added 5205b19 because in my work on my next PR, i realized that the `next` methods should return null when the *next* node is the termination entry in the list - similar to how iterating over a null-terminated slice doesn't yield the null terminator in the final iteration.

linusg commented:
Apologies for the delay on this, I haven't forgotten about it yet 🙂

linusg reviewed (changes requested):
Finally got around to testing this with my own project - std.debug needs an update:

```diff
--- a/lib/std/debug.zig
+++ b/lib/std/debug.zig
@@ -629,9 +629,9 @@ pub fn defaultPanic(
             // isn't visible on actual hardware if directly booted into
             inline for ([_]?*uefi.protocol.SimpleTextOutput{ uefi.system_table.std_err, uefi.system_table.con_out }) |o| {
                 if (o) |out| {
-                    _ = out.setAttribute(uefi.protocol.SimpleTextOutput.red);
-                    _ = out.outputString(exit_msg);
-                    _ = out.setAttribute(uefi.protocol.SimpleTextOutput.white);
+                    out.setAttribute(.{ .foreground = .red }) catch {};
+                    _ = out.outputString(exit_msg) catch {};
+                    out.setAttribute(.{ .foreground = .white }) catch {};
                 }
             }
```

Please do a quick grep around std to see if any other UEFI-specific code paths need updating.

---

Once this is merged I think we should look into supporting returning `std.os.uefi.Status.Error` from `main()`, having to convert the errors back into a status every time is a bit tedious :)

dotcarmen commented:
I grepped and none of the other `uefi` symbols in `lib/std` are using these APIs 😅 they're attempting to use boot_services or runtime_services (which this PR doesn't affect) or checking the current os tag for other usage (path separaters, string selection).

none of the rest of the codebase uses anything from `std.os.uefi` :)

good idea! i'll open a PR in a few :)
